apps/ejabberd/src/mod_mam_meta.erl

```diff
@@ -0,0 +1,177 @@
+    DepsWithPmAndMuc = handle_nested_opts(muc, Opts, false, DepsWithPm),
+
+    [{Dep, Args, hard} || {Dep, Args} <- maps:to_list(DepsWithPmAndMuc)].
+
+%%--------------------------------------------------------------------
+%% Helpers
+%%--------------------------------------------------------------------
+
+-spec handle_nested_opts(Key :: atom(), RootOpts :: proplists:proplist(),
+                         Default :: term(), deps()) -> deps().
+handle_nested_opts(Key, RootOpts, Default, Deps) ->
+    case proplists:get_value(Key, RootOpts, Default) of
+        false -> Deps;
+        Opts0 ->
+            Opts = normalize(Opts0),
+            FullOpts = lists:ukeymerge(1, Opts, RootOpts),
+            parse_opts(Key, FullOpts, Deps)
+    end.
+
+
+-spec parse_opts(Type :: pm | muc, Opts :: proplists:proplist(), deps()) -> deps().
+parse_opts(Type, Opts, Deps) ->
+    CoreMod =
+        case Type of
+            pm -> mod_mam;
+            muc -> mod_mam_muc
+        end,
+
+    CoreModOpts =
+        lists:filtermap(
+          fun(Key) ->
+                  case proplists:lookup(Key, Opts) of
+                      none -> false;
+                      Opt -> {true, Opt}
+                  end
+          end,
+          [add_archived_element, is_complete_message, host]),
+
+    WithCoreDeps = add_dep(CoreMod, CoreModOpts, Deps),
+    Backend = proplists:get_value(backend, Opts, odbc),
+
+    parse_backend_opts(Backend, Type, Opts, WithCoreDeps).
+
+
+-spec parse_backend_opts(odbc | cassandra | riak, Type :: pm | muc,
+                         Opts :: proplists:proplist(), deps()) -> deps().
+parse_backend_opts(cassandra, Type, Opts, Deps0) ->
+    ModArch =
+        case Type of
+            pm -> mod_mam_cassandra_arch;
+            muc -> mod_mam_muc_cassandra_arch
+        end,
+
+    Deps = add_dep(ModArch, Deps0),
+
+    case proplists:get_value(user_prefs_store, Opts, false) of
+        cassandra -> add_dep(mod_mam_cassandra_prefs, [Type], Deps);
+        mnesia -> add_dep(mod_mam_mnesia_prefs, [Type], Deps);
+        mnesia_dirty -> add_dep(mod_mam_mnesia_dirty_prefs, [Type], Deps);
+        _ -> Deps
+    end;
+
+parse_backend_opts(riak, Type, Opts, Deps0) ->
+    Deps = add_dep(mod_mam_riak_timed_arch_yz, [Type], Deps0),
+
+    case proplists:get_value(user_prefs_store, Opts, false) of
+        mnesia -> add_dep(mod_mam_mnesia_prefs, [Type], Deps);
+        mnesia_dirty -> add_dep(mod_mam_mnesia_dirty_prefs, [Type], Deps);
+        _ -> Deps
+    end;
+
+parse_backend_opts(odbc, Type, Opts0, Deps0) ->
+    Opts = add_default_odbc_opts(Opts0),
+
+    {ModODBCArch, ModAsyncWriter} =
+        case Type of
+            pm -> {mod_mam_odbc_arch, mod_mam_odbc_async_pool_writer};
+            muc -> {mod_mam_muc_odbc_arch, mod_mam_muc_odbc_async_pool_writer}
+        end,
+
+    Deps1 = add_dep(ModODBCArch, [Type], Deps0),
+    Deps = add_dep(mod_mam_odbc_user, [Type], Deps1),
+
+    lists:foldl(
+      fun
```

Comment:
Extracting this `fun` to a function will most probably improve readability here.